Unify meta tensor and fake tensor converter conversion

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87943
* #87911
* #87903

Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @jansel @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

ezyang committed:
Unify meta tensor and fake tensor converter conversion

Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -280,10 +295,13 @@ def __call__(self, t, shape_env=None):
+                # TODO: this is suspicious, now that we have callback argument
                 if type(t) is torch.nn.Parameter:
                     r = torch.nn.Parameter(r, requires_grad=r.requires_grad)
                 return r
```
Should we move this logic https://github.com/pytorch/pytorch/blob/48e1981d0286c75afe043750c3f3b4f6b841405c/torch/_subclasses/fake_tensor.py#L235-L239 here ?

ezyang replied:
Oh yeah, good catch

ezyang replied:
I moved it into meta_tensor itself, instead of at top level. This way recursive grad field also works.

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -127,18 +127,24 @@ def del_ten():
-            self.storage_memo[swr] = torch.empty(s.size(), dtype=s.dtype, device="meta")
+            self.storage_memo[swr] = (
+                callback(
+                    lambda: torch.empty(s.size(), dtype=torch.uint8, device="meta")
```
`s.dtype ->torch.uint8` - don't we need `s.size() * s.element_size()` if we're going to use `uint8`

ezyang replied:
No because it's untyped storage so already reporting size in bytes. Maybe could use nbytes() to ne clearer.

eellison reviewed (commented):
Cool! just asking some questions and want to make sure I understand. 

> I can't store tensors for the storages anymore, as that will result
in a leak. But we have untyped storage now, so I just store untyped
storages instead.

Why is this ?

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -229,11 +242,13 @@ def is_c_of_r(complex_dtype, real_dtype):
                                 # sort of unsupported grad_fn here
                                 r = r.clone()
```
I changed this: fake tensor mode *is* active here. So r will be fake

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -229,11 +242,13 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        assert safe_is_leaf(
+                            r
+                        ), "the callback you passed in doesn't detach"
+                    s = self.meta_storage(t.storage().untyped(), callback=callback)
```
Without the callback, fake tensor mode is active and I attempt to create a meta fake tensor which is not allowed by default. So I must use the callback to instead make a non meta storage (which actually is meta cuz fake mode).

ezyang committed:
Update on "Unify meta tensor and fake tensor converter conversion"


Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
Ran into a funny bug in autograd https://github.com/pytorch/pytorch/issues/87956 so I had to refactor meta converter a bit more to make it better.

ezyang commented:
> Why is this ?

The storage map is a strong map and I didn't feel like changing this

ezyang committed:
Update on "Unify meta tensor and fake tensor converter conversion"


Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc jansel mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ezyang committed:
Update on "Unify meta tensor and fake tensor converter conversion"


Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc jansel mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ezyang commented:
So there was another subtle bug, where `no_dispatch` was on (due to fake tensor enabling it) when I attempt to clone in converter (to fake up autograd history). To fix this, I added an assert to converter that no dispatch is NOT disabled (invariant: you must always dispatch), and then refactored fake tensor to push the no dispatch / in kernel manager calls as close to the call sites as possible.

Once I did this, I also found that I had to setup in kernel manager for `set_` call. I break encapsulation and just do this by hand.

ezyang commented:
If you desire a smaller patch, I could probably extract out the "make no dispatch extents smaller" into its own refactor patch, but I'd prefer not to.

ezyang committed:
Update on "Unify meta tensor and fake tensor converter conversion"


Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc jansel mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ezyang committed:
Update on "Unify meta tensor and fake tensor converter conversion"


Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc jansel mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

eellison commented:
Would it be worth considering adding an API to re-enable torch dispatch, or an api to disable it such that it could be re-enabled ? I think that might make the changes easier/less invasive. For instance, instead of using 'no_dispatch' directly, we could add the a context manager like in_kernel_invocation that adds the torch._C.DisableTorchDispatch object on the fake tensor mode itself such that a subsequent context manager could temporarily delete it

ezyang commented:
I think that the newly refactored code is better. Previously, it seemed that we were indiscriminately spraying no_dispatch() inside the inner implementation of fake tensor. no_dispatch without in kernel mode is dangerous because any op you call can cause a segfault, as the fake tensor is still advertising that its a CPU tensor. It is better to keep the no_dispatch calls scoped as tightly as possible.

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -655,6 +662,13 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
 
+        # Some attribute queries that can be serviced directly
+        # See Note [is_coalesced is dispatched]
+        if func in [torch.ops.aten.is_coalesced.default]:
```
reason not to register this with `register_op_impl` ?

ezyang replied:
It needs to happen early. In particular, it needs to happen before we check for non-fake, as the triggering case was we are in fake tensor mode querying if a non fake tensor is coalesced.

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        and r.storage_offset() == storage_offset
+                    ):
+                        # You're normal and happy, install the fresh storage into the memo
+                        self.storage_memo[swr] = r.storage().untyped()
```
seems a bit strange we have to directly invoke the memo here instead of as part of normal conversion

ezyang replied:
The old code always created a storage, and then assembled it. This new path tries to create the tensors in a more "logical" way, which is using the normal `empty_strided` constructor. However, this constructor goes ahead and creates a storage for you, so you shouldn't try to create one yourself if you're OK with this storage.

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
         return self.get_tensor_memo(t)
 
-    def __call__(self, t, shape_env=None):
+    def __call__(self, t, shape_env=None, *, strict=False, callback=lambda t: t()):
```
the only other callee of this is `run_meta_crossref` - could remove the `strict=False` arg, always return NotImplemented, and remove this arg. not a big deal either way

ezyang replied:
I was going to say NotImplemented is annoying to deal with in run_meta_crossref because we run the converter inside of a `tree_map`, but I guess if we return non-Tensor inputs without returning NotImplemented there it is fine, I can still query the miss variable.

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
+                    warnings.filterwarnings("ignore", "The .grad attribute of a Tensor")
+                    grad_not_none = t.grad is not None
+                if grad_not_none:
+                    r.grad = self.meta_tensor(t.grad, shape_env, callback)
```
What happens if the `grad` here isn't convertible to meta ? is that possible/something we need to handle ?

ezyang replied:
not sure how it wouldn't be convertible. Seems like the same thing as asking what if r is not convertible. We... fix it.

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -280,10 +361,13 @@ def __call__(self, t, shape_env=None):
-                r = self.meta_tensor(t, shape_env=shape_env)
+                r = self.meta_tensor(t, shape_env=shape_env, callback=callback)
+                # TODO: this is suspicious, now that we have callback argument
                 if type(t) is torch.nn.Parameter:
```
this is suspicious because the callback could call `torch.nn.Parameter` on the output instead of here ?

ezyang replied:
yeah, the whole point of the callback argument was to make sure we preserved non-leafness, but now we've gone and wrapped it in another Tensor constructor after we processed everything. But maybe it's ok, cuz parameters are always leafs anyway??

eellison reviewed (approved):
🚀

ezyang committed:
Update on "Unify meta tensor and fake tensor converter conversion"


Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc jansel mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

albanD commented on torch/_subclasses/meta_utils.py:
```diff
@@ -127,18 +129,31 @@ def del_ten():
+        # This indicates you set no_dispatch() before calling into this
+        # function.  This is an error: we may be creating fake tensors and
+        # will perform operations on them which need fake tensor mode to
+        # be active.  You will segfault if you are in a no_dispatch() block.
```
Can we actually fix that? You can very easily segfault if you use no_dispatch. That is ~expected.

ezyang replied:
Yeah, it's not possible to fix, I think. The problem is that `no_dispatch` will strip the Python key, but if you don't add Meta to the include set, you'll just go straight to a CPU kernel with a fake tensor, and then all bets are off. There is no logical interposition point at this point.

albanD commented on torch/_subclasses/meta_utils.py:
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        assert safe_is_leaf(
+                            r
+                        ), "the callback you passed in doesn't detach"
+                        r.requires_grad = t.requires_grad
```
Now that this is moved outside of the empty call, the logic here is:
```
if ..:
  do A
  do B
else:
  do A
```
We can clean that up.

albanD commented on torch/_subclasses/meta_utils.py:
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        self.storage_memo[swr] = r.storage().untyped()
+                    else:
+                        # You're in crazy town; somehow you gave us a tensor
+                        # that wasn't a view, but had nonzero storage offset,
```
nit: wasn't a view in what sense? "differentiable view" ?
If so, you can easily get here by detach-ing any weird Tensor you had.

ezyang replied:
Hmm, you are right! Do you think it is better if I view and detach in this case, instead of using set_?

albanD reviewed (approved):
Sounds good!

ezyang commented on test/test_meta.py:
```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "The .grad attribute of a Tensor")
+            grad_not_none = m1.grad is not None
+        if grad_not_none:
```
yes, good catch

ezyang commented on test/test_meta.py:
```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
+        if m1.is_sparse:
+            self.assertEqual(m1.dense_dim(), m2.dense_dim())
+            self.assertEqual(m1.sparse_dim(), m2.sparse_dim())
+            self.assertEqual(m1.is_coalesced(), m2.is_coalesced())
```
yes, I suppose we should

ezyang replied:
Actually, this would fail, because I don't propagate nnz when fakifying. nnz is very data dependent, so it's not clear it's worth propagating.

ezyang commented on test/test_meta.py:
```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
+            if m1._is_view():
+                self.assertMetadataMatches(m1._base, m2._base)
+        # TODO: test if is resizable (no direct query for this atm)
+        # TODO: audit AutogradMeta to see if it matches
```
Forward AD agreed.

Hooks, I'm not so sure. Can a hook trigger if it is attached to the INPUT tensor of a grad compute? I guess if it's a pre-hook? But it will trigger anyway when the gradient flows to the original, real tensor. So you might double trigger the hook in that situation.

ezyang commented:
@pytorchbot merge

ezyang commented:
Gonna handle comments on follow up

kit1980 commented:
@pytorchbot revert -m "Broke several inductor tests" -c nosignal
